File: examples/go/ordinal-nft/OrdinalNFT_test.go
package contract

import (
	"encoding/hex"
	"encoding/json"
	"strings"
	"testing"

	"github.com/icellan/runar/compilers/go/compiler"
	runar "github.com/icellan/runar/packages/runar-go"
)

// ---------------------------------------------------------------------------
// Shared helpers
// ---------------------------------------------------------------------------

func compileContract(t *testing.T) *runar.RunarArtifact {
	t.Helper()
	a, err := compiler.CompileFromSource("OrdinalNFT.runar.go")
	if err != nil {
		t.Fatalf("OrdinalNFT compile failed: %v", err)
	}
	raw, err := compiler.ArtifactToJSON(a)
	if err != nil {
		t.Fatalf("artifact marshal failed: %v", err)
	}
	var out runar.RunarArtifact
	if err := json.Unmarshal(raw, &out); err != nil {
		t.Fatalf("artifact unmarshal failed: %v", err)
	}
	return &out
}

// ---------------------------------------------------------------------------
// Business logic tests
// ---------------------------------------------------------------------------

func TestOrdinalNFT_Unlock(t *testing.T) {
	pk := runar.Alice.PubKey
	c := &OrdinalNFT{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.SignTestMessage(runar.Alice.PrivKey), pk)
}

func TestOrdinalNFT_Unlock_WrongKey(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("expected assertion failure for wrong public key")
		}
	}()
	pk := runar.Alice.PubKey
	c := &OrdinalNFT{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.SignTestMessage(runar.Bob.PrivKey), runar.Bob.PubKey)
}

// TestOrdinalNFT_IsStateless verifies the contract has no mutable state —
// readonly properties stay in place after Unlock.
func TestOrdinalNFT_IsStateless(t *testing.T) {
	pk := runar.Alice.PubKey
	c := &OrdinalNFT{PubKeyHash: runar.Hash160(pk)}
	c.Unlock(runar.SignTestMessage(runar.Alice.PrivKey), pk)
	if len(c.PubKeyHash) == 0 {
		t.Error("expected PubKeyHash to remain non-empty after unlock")
	}
}

// ---------------------------------------------------------------------------
// Rúnar compile check
// ---------------------------------------------------------------------------

func TestOrdinalNFT_Compile(t *testing.T) {
	if err := runar.CompileCheck("OrdinalNFT.runar.go"); err != nil {
		t.Fatalf("Rúnar compile check failed: %v", err)
	}
}

func TestOrdinalNFT_CompilesToValidArtifact(t *testing.T) {
	a := compileContract(t)
	if a.ContractName != "OrdinalNFT" {
		t.Errorf("expected contractName OrdinalNFT, got %q", a.ContractName)
	}
	if len(a.ABI.Methods) != 1 {
		t.Fatalf("expected 1 method, got %d", len(a.ABI.Methods))
	}
	if a.ABI.Methods[0].Name != "unlock" {
		t.Errorf("expected method 'unlock', got %q", a.ABI.Methods[0].Name)
	}
}

// ---------------------------------------------------------------------------
// SDK inscription flow — image/png
// ---------------------------------------------------------------------------

func TestOrdinalNFT_AttachesImageInscription(t *testing.T) {
	a := compileContract(t)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})

	// Small PNG header as inscription data
	pngData := "89504e470d0a1a0a"
	contract.WithInscription(&runar.Inscription{ContentType: "image/png", Data: pngData})

	lockingScript := contract.GetLockingScript()

	expectedEnvelope := runar.BuildInscriptionEnvelope("image/png", pngData)
	if !strings.Contains(lockingScript, expectedEnvelope) {
		t.Errorf("locking script does not contain expected envelope\n  script: %s\n  envelope: %s",
			lockingScript, expectedEnvelope)
	}

	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	if parsed.ContentType != "image/png" {
		t.Errorf("contentType = %q, want image/png", parsed.ContentType)
	}
	if parsed.Data != pngData {
		t.Errorf("data = %q, want %q", parsed.Data, pngData)
	}
}

// ---------------------------------------------------------------------------
// SDK inscription flow — text/plain
// ---------------------------------------------------------------------------

func TestOrdinalNFT_AttachesTextInscription(t *testing.T) {
	a := compileContract(t)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})

	textData := hex.EncodeToString([]byte("Hello, Ordinals!"))
	contract.WithInscription(&runar.Inscription{ContentType: "text/plain", Data: textData})

	lockingScript := contract.GetLockingScript()

	parsed := runar.ParseInscriptionEnvelope(lockingScript)
	if parsed == nil {
		t.Fatal("expected inscription envelope, got nil")
	}
	if parsed.ContentType != "text/plain" {
		t.Errorf("contentType = %q, want text/plain", parsed.ContentType)
	}
	decoded, err := hex.DecodeString(parsed.Data)
	if err != nil {
		t.Fatalf("decode text data: %v", err)
	}
	if string(decoded) != "Hello, Ordinals!" {
		t.Errorf("decoded text = %q, want %q", string(decoded), "Hello, Ordinals!")
	}
}

// ---------------------------------------------------------------------------
// Round-trip through FromUtxo
// ---------------------------------------------------------------------------

func TestOrdinalNFT_InscriptionRoundTripFromUtxo(t *testing.T) {
	a := compileContract(t)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})

	pngData := "89504e470d0a1a0a"
	contract.WithInscription(&runar.Inscription{ContentType: "image/png", Data: pngData})

	lockingScript := contract.GetLockingScript()

	reconnected := runar.FromUtxo(a, runar.UTXO{
		Txid:        "0000000000000000000000000000000000000000000000000000000000000000",
		OutputIndex: 0,
		Satoshis:    1,
		Script:      lockingScript,
	})
	insc := reconnected.GetInscription()
	if insc == nil {
		t.Fatal("expected inscription after FromUtxo")
	}
	if insc.ContentType != "image/png" {
		t.Errorf("contentType = %q, want image/png", insc.ContentType)
	}
	if insc.Data != pngData {
		t.Errorf("data = %q, want %q", insc.Data, pngData)
	}
}

// ---------------------------------------------------------------------------
// Locking script without inscription has no envelope
// ---------------------------------------------------------------------------

func TestOrdinalNFT_LockingScriptWithoutInscription(t *testing.T) {
	a := compileContract(t)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})

	lockingScript := contract.GetLockingScript()
	if parsed := runar.ParseInscriptionEnvelope(lockingScript); parsed != nil {
		t.Errorf("expected no inscription envelope, got %+v", parsed)
	}
}

// ---------------------------------------------------------------------------
// WithInscription returns receiver for chaining
// ---------------------------------------------------------------------------

func TestOrdinalNFT_WithInscriptionChaining(t *testing.T) {
	a := compileContract(t)
	contract := runar.NewRunarContract(a, []interface{}{runar.Alice.PubKeyHash})
	result := contract.WithInscription(&runar.Inscription{ContentType: "text/plain", Data: ""})
	if result != contract {
		t.Error("WithInscription should return the same contract for chaining")
	}
}

File: examples/sol/ordinal-nft/OrdinalNFT.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { TestContract, ALICE, BOB, signTestMessage } from 'runar-testing';
import { compile } from 'runar-compiler';
import type { RunarArtifact } from 'runar-ir-schema';
import { RunarContract } from '../../../packages/runar-sdk/src/contract.js';
import {
  buildInscriptionEnvelope,
  parseInscriptionEnvelope,
} from '../../../packages/runar-sdk/src/ordinals/envelope.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const source = readFileSync(join(__dirname, 'OrdinalNFT.runar.sol'), 'utf8');
const FILE_NAME = 'OrdinalNFT.runar.sol';

const SIG = signTestMessage(ALICE.privKey);
const BOB_SIG = signTestMessage(BOB.privKey);

function utf8ToHex(str: string): string {
  return Array.from(new TextEncoder().encode(str))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

function hexToUtf8(hex: string): string {
  const bytes = new Uint8Array(hex.length / 2);
  for (let i = 0; i < hex.length; i += 2) {
    bytes[i / 2] = parseInt(hex.slice(i, i + 2), 16);
  }
  return new TextDecoder().decode(bytes);
}

function compileContract(): RunarArtifact {
  const result = compile(source, { fileName: FILE_NAME });
  if (!result.success || !result.artifact) {
    const errors = result.diagnostics
      .filter((d) => d.severity === 'error')
      .map((d) => d.message)
      .join('\n  ');
    throw new Error(`OrdinalNFT compile failed:\n  ${errors}`);
  }
  return result.artifact;
}

describe('OrdinalNFT (Solidity)', () => {
  describe('business logic', () => {
    it('accepts a valid unlock with correct pubkey and signature', () => {
      const contract = TestContract.fromSource(source, { pubKeyHash: ALICE.pubKeyHash }, FILE_NAME);
      const result = contract.call('unlock', { sig: SIG, pubKey: ALICE.pubKey });
      expect(result.success).toBe(true);
    });

    it('rejects unlock with wrong public key', () => {
      const contract = TestContract.fromSource(source, { pubKeyHash: ALICE.pubKeyHash }, FILE_NAME);
      const result = contract.call('unlock', { sig: BOB_SIG, pubKey: BOB.pubKey });
      expect(result.success).toBe(false);
    });
  });

  describe('compilation', () => {
    it('compiles to a valid artifact', () => {
      const artifact = compileContract();
      expect(artifact.contractName).toBe('OrdinalNFT');
      expect(artifact.abi.methods).toHaveLength(1);
      expect(artifact.abi.methods[0]!.name).toBe('unlock');
    });
  });

  describe('SDK inscription flow', () => {
    it('attaches an image inscription to the locking script', () => {
      const artifact = compileContract();
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);

      const pngData = '89504e470d0a1a0a';
      contract.withInscription({ contentType: 'image/png', data: pngData });

      const lockingScript = contract.getLockingScript();
      const expectedEnvelope = buildInscriptionEnvelope('image/png', pngData);
      expect(lockingScript).toContain(expectedEnvelope);

      const parsed = parseInscriptionEnvelope(lockingScript);
      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('image/png');
      expect(parsed!.data).toBe(pngData);
    });

    it('attaches a text inscription to the locking script', () => {
      const artifact = compileContract();
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);

      const textData = utf8ToHex('Hello, Ordinals!');
      contract.withInscription({ contentType: 'text/plain', data: textData });

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);
      expect(parsed).not.toBeNull();
      expect(parsed!.contentType).toBe('text/plain');
      expect(hexToUtf8(parsed!.data)).toBe('Hello, Ordinals!');
    });

    it('round-trips inscription through fromUtxo', () => {
      const artifact = compileContract();
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);

      const pngData = '89504e470d0a1a0a';
      contract.withInscription({ contentType: 'image/png', data: pngData });

      const lockingScript = contract.getLockingScript();

      const reconnected = RunarContract.fromUtxo(artifact, {
        txid: '00'.repeat(32),
        outputIndex: 0,
        satoshis: 1,
        script: lockingScript,
      });

      expect(reconnected.inscription).not.toBeNull();
      expect(reconnected.inscription!.contentType).toBe('image/png');
      expect(reconnected.inscription!.data).toBe(pngData);
    });

    it('locking script without inscription has no envelope', () => {
      const artifact = compileContract();
      const contract = new RunarContract(artifact, [ALICE.pubKeyHash]);

      const lockingScript = contract.getLockingScript();
      const parsed = parseInscriptionEnvelope(lockingScript);
      expect(parsed).toBeNull();
    });
  });
});
